### PR TITLE
update samples to use migration v2

### PR DESCRIPTION
**Pull request by w-henderson**
This commit updates the code samples to demonstrate usage of the migration v2 API.

### CODE REVIEW

File: samples/client.c [hunk 1]
```diff
@@ -688,42 +793,41 @@ handle_dma_io(int sock, struct vfio_user_dma_map *dma_regions,
...
+    res->flags = VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT
+               | VFIO_DEVICE_FEATURE_GET;
+    res->argsz = size;
 
-    range = data + sizeof(*dirty_pages);
-    range->iova = dma_region->addr;
-    range->size = dma_region->size;
-    range->bitmap.size = bitmap_size;
-    range->bitmap.pgsize = sysconf(_SC_PAGESIZE);
+    report = data + sizeof(*res);
```
Comment (tmakatos) [inline samples/client.c:814]: nit: `res + 1;`

File: samples/client.c [hunk 2]
```diff
@@ -884,16 +961,15 @@ migrate_from(int sock, size_t *nr_iters, struct iovec **migr_iters,
...
+    device_state = VFIO_USER_DEVICE_STATE_PRE_COPY;
+    ret = set_migration_state(sock, device_state);
     if (ret < 0) {
         err(EXIT_FAILURE, "failed to write to device state");
     }
 
-    _nr_iters = do_migrate(sock, 1, *migr_iters);
-    assert(_nr_iters == 1);
+    iters = do_migrate(sock, *nr_iters, max_iter_size, *migr_iters);
+    assert(iters == 12);
```
Comment (tmakatos) [inline samples/client.c:971]: Where does 12 come from?

File: samples/client.c [hunk 3]
```diff
@@ -905,31 +981,23 @@ migrate_from(int sock, size_t *nr_iters, struct iovec **migr_iters,
...
     }
 
-    _nr_iters += do_migrate(sock, 1, (*migr_iters) + _nr_iters);
-    if (_nr_iters != 2) {
-        errx(EXIT_FAILURE,
-             "expected 2 iterations instead of %zu while in stop-and-copy state",
-             _nr_iters);
-    }
+    iters = do_migrate(sock, *nr_iters, max_iter_size, *migr_iters);
+    assert(iters == 13);
```
Comment (tmakatos) [inline samples/client.c:991]: Where does 13 come from?

File: samples/client.c [hunk 4]
```diff
@@ -874,7 +946,12 @@ migrate_from(int sock, size_t *nr_iters, struct iovec **migr_iters,
         err(EXIT_FAILURE, "failed to create pthread");
     }
 
-    *nr_iters = 2;
+    size_t expected_data = bar1_size + sizeof(time_t);
+
+    *nr_iters = (expected_data + max_iter_size - 1) / max_iter_size;
+
+    printf("client: calculated %ld iters needed to migrate\n", *nr_iters);
```
Comment (tmakatos) [inline samples/client.c:953]: Not sure this helps, if we know the size we can add an assert?

File: samples/client.c [hunk 5]
```diff
@@ -992,57 +1060,29 @@ migrate_to(char *old_sock_path, int *server_max_fds,
...
-         * TODO write half of migration data via regular write and other half via
-         * memopy map.
-         */
-        printf("client: writing migration device data %#llx-%#llx\n",
-               (ull_t)data_offset,
-               (ull_t)(data_offset + migr_iters[i].iov_len - 1));
-        ret = access_region(sock, VFU_PCI_DEV_MIGR_REGION_IDX, true,
-                            data_offset, migr_iters[i].iov_base,
-                            migr_iters[i].iov_len);
+        printf("client: writing migration device data iter %zu\n", i);
```
Comment (tmakatos) [inline samples/client.c:1073]: nit: not sure how much this print helps

File: samples/client.c [hunk 6]
```diff
@@ -1056,6 +1096,15 @@ migrate_to(char *old_sock_path, int *server_max_fds,
     if (dst_crc != src_crc) {
         fprintf(stderr, "client: CRC mismatch: %u != %u\n", src_crc, dst_crc);
         abort();
+    } else {
+        printf("client: CRC match, we did it! :)\n");
```
Comment (tmakatos) [inline samples/client.c:1100]: nit: not sure how much this helps given we abort if it doesn't match

File: samples/client.c [hunk 7]
```diff
@@ -1108,6 +1156,14 @@ int main(int argc, char *argv[])
     uint32_t crc;
     size_t bar1_size = 0x3000; /* FIXME get this value from region info */
 
+    struct vfio_user_device_feature *dirty_pages_feature;
+    struct vfio_user_device_feature_dma_logging_control *dirty_pages_control;
+    size_t dirty_pages_size = sizeof(*dirty_pages_feature) +
+                               sizeof(*dirty_pages_control);
+    void *dirty_pages = malloc(dirty_pages_size);
+    dirty_pages_feature = dirty_pages;
+    dirty_pages_control = dirty_pages + sizeof(*dirty_pages_feature);
```
Comment (tmakatos) [inline samples/client.c:1165]: `dirty_pages_feature + 1`

File: samples/client.c [hunk 8]
```diff
@@ -1202,11 +1258,15 @@ int main(int argc, char *argv[])
      */
     irq_fd = configure_irqs(sock);
 
-    dirty_pages.argsz = sizeof(dirty_pages);
-    dirty_pages.flags = VFIO_IOMMU_DIRTY_PAGES_FLAG_START;
-    ret = tran_sock_msg(sock, 0, VFIO_USER_DIRTY_PAGES,
-                        &dirty_pages, sizeof(dirty_pages),
-                        NULL, NULL, 0);
+    dirty_pages_feature->argsz = sizeof(*dirty_pages_feature) +
```
Comment (tmakatos) [inline samples/client.c:1261]: add comment before this line `/* start logging dirty pages */` (or something)

File: samples/client.c [hunk 9]
```diff
@@ -1230,11 +1290,15 @@ int main(int argc, char *argv[])
         get_dirty_bitmap(sock, &dma_regions[i]);
     }
 
-    dirty_pages.argsz = sizeof(dirty_pages);
-    dirty_pages.flags = VFIO_IOMMU_DIRTY_PAGES_FLAG_STOP;
-    ret = tran_sock_msg(sock, 0, VFIO_USER_DIRTY_PAGES,
-                        &dirty_pages, sizeof(dirty_pages),
-                        NULL, NULL, 0);
+    dirty_pages_feature->argsz = sizeof(*dirty_pages_feature) +
```
Comment (tmakatos) [inline samples/client.c:1293]: add comment before this line /* stop logging dirty pages */ (or something)

File: samples/client.c [hunk 10]
```diff
@@ -1271,7 +1335,9 @@ int main(int argc, char *argv[])
         err(EXIT_FAILURE, "failed to write to BAR0");
     }
 
-    nr_iters = migrate_from(sock, &nr_iters, &migr_iters, &crc, bar1_size);
+    nr_iters = migrate_from(sock, &nr_iters, &migr_iters, &crc, bar1_size,
+                            MIN(server_max_data_xfer_size,
+                                CLIENT_MAX_DATA_XFER_SIZE));
```
Comment (tmakatos) [inline samples/client.c:1340]: identation

File: samples/server.c [hunk 11]
```diff
@@ -285,125 +286,112 @@ migration_device_state_transition(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state)
...
      * both BAR1 and BAR0. Since we always copy BAR1 in the stop-and-copy state,
      * copying BAR1 in the pre-copy state is pointless. Fixing this requires
      * more complex state tracking which exceeds the scope of this sample.
      */
 
-    if (offset != 0 || size != server_data->migration.pending_bytes) {
-        errno = EINVAL;
-        return -1;
-    }
+    uint32_t total_read = server_data->bar1_size;
```
Comment (tmakatos) [inline samples/server.c:301]: Is `total_read` remaining or read (past simple)?

File: samples/server.c [hunk 12]
```diff
@@ -285,125 +286,112 @@ migration_device_state_transition(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state)
...
+        total_read += sizeof(server_data->bar0);
+    }
+
+    if (server_data->migration.bytes_transferred == total_read || size == 0) {
+        vfu_log(vfu_ctx, LOG_DEBUG, "no data left to read");
+        return 0;
+    }
+
+    uint32_t read_start = server_data->migration.bytes_transferred;
+    uint32_t read_end = MIN(read_start + size, total_read); // exclusive
```
Comment (tmakatos) [inline samples/server.c:313]: "exclusive"?

File: samples/server.c [hunk 13]
```diff
@@ -285,125 +286,112 @@ migration_device_state_transition(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state)
...
+    assert(read_end > read_start);
+
+    uint32_t bytes_read = read_end - read_start;
+
+    if (read_end <= server_data->bar1_size) {
```
Comment (tmakatos) [inline samples/server.c:318]: Isn't is simpler to do:
```
int bar0_size = 0;
/* read BAR0, if any */
if (read_start < server_data->bar0_size) {
    bar0_size = MIN(size, server_data->bar0_size - read_start);
    memcpy(buf, server_data->bar0 + read_start, bar0_size);
    size -= bar0_size;
    read_start += size;
}
/* read BAR1 */
if (read_end >= server_data->bar0_size) {
    bar1_size = MIN(read_end, server_data->bar0_size);
    memcpy(buf + bar0_size, server_data->bar0 + read_start, bar1_size);
}
```
That's just a skectch, but you get the idea.

File: samples/server.c [hunk 14]
```diff
@@ -285,125 +286,112 @@ migration_device_state_transition(vfu_ctx_t *vfu_ctx, vfu_migr_state_t state)
...
-static int
-migration_data_written(UNUSED vfu_ctx_t *vfu_ctx, UNUSED uint64_t count)
-{
-    /*
-     * We apply migration state directly in the migration_write_data callback,
-     * so we don't need to do anything here. We would have to apply migration
-     * state in this callback if the migration region was memory mappable, in
-     * which case we wouldn't know when the client wrote migration data.
-     */
+    if (write_end <= server_data->bar1_size) {
```
Comment (tmakatos) [inline samples/server.c:370]: Same comment as in `migration_read_data`.

File: samples/client.c (hunk 3 above)
Comment (w-henderson) [reply samples/client.c:991]: The same as above, but we also have 8 bytes for `time_t` so it takes 13 iterations. I'll have this compare to an updated `*nr_iters` too.

File: samples/client.c (hunk 5 above)
Comment (w-henderson) [reply samples/client.c:1073]: Sure, I'll get rid of it.

File: samples/server.c (hunk 11 above)
Comment (w-henderson) [reply samples/server.c:301]: The total amount we are going to have to read. The amount we have already read is `server_data->migration.bytes_transferred`. Do you think I should rename this variable, and if so, what to?

Comment (tmakatos) [reply samples/server.c:301]: I think `remaining` is better?

Comment (w-henderson) [reply samples/server.c:301]: I'm not sure, it's not really "remaining", it's more like "the size of everything we have to transfer, regardless of how much we've already done"...?

Comment (tmakatos) [reply samples/server.c:301]: `total_to_read` ?

**Review by tmakatos — APPROVED** [review]
LGTM, re: `total_read`, you can just add a comment instead of renaming the variable.

File: samples/server.c (hunk 11 above)
Comment (w-henderson) [reply samples/server.c]: Sounds good.

**Review by tmakatos — APPROVED** [review]

**Pull request merged**